Support virtualCapacities in nodeTemplates

Maydaer opened the pull request:
**How to categorize this PR?**

/area auto-scaling
/kind enhancement
/platform aws

**What this PR does / why we need it**:
- removes requirement that nodeTemplate.Capacities have to be non-empty, since `nodeTemplate.virtualCapacity` may be set
- adds validation for extended resources (only in `nodetemplate.VirtualCapacity`, regular capacity validation is kept as they are for backward compatibility)

**Which issue(s) this PR fixes**:
Fixes #1238

**Special notes for your reviewer**:
Nothing

**Release note**:
```other operator
nodeTemplate.Capacity is now allowed to be empty. Extended resources, memory and GPU are now validated as whole and non-negative numbers.
```

gardener-robot-ci-3 commented:
Thank you @Maydaer for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

kon-angelo commented on pkg/apis/aws/validation/worker.go:
```diff
@@ -71,17 +72,18 @@ func ValidateWorkerConfig(workerConfig *apisaws.WorkerConfig, volume *core.Volum
+			if capacityAttribute == corev1.ResourceCPU { // CPU is allowed to be a decimal value, such as 1500m.
+				allErrs = append(allErrs, validateResourceQuantityValue(capacityAttribute, value, fldPath.Child("nodeTemplate").Child("capacity").Child(string(capacityAttribute)))...)
+			} else { // extended resources need to be whole, non-negative numbers. GPU and memory also need to be whole numbers.
+				allErrs = append(allErrs, validateResourceQuantityWholeNumber(capacityAttribute, value, fldPath.Child("nodeTemplate").Child("capacity").Child(string(capacityAttribute)))...)
```
I would really appreciate a comment with the link to the ext.resources page because for sure we will forget the integer only restriction. 

Also I am not sure if there is any point in advertising a resource of `0` value. Is it something that we should maybe validate against ?

Maydaer commented:
fixing typos

Maydaer commented:
Hi @aaronfern
1. I don't see an important use case for allowing an empty `nodeTemplate`, but I also do think it is in principle valid to have an empty `nodeTemplate`. The behavior for an empty nodeTemplate is well defined. I think you can argue that the user might have forgotten to set a capacity or virtualCapacity if both are empty, but the user might also have intentionally chosen an implementation that has the potential to enter an empty nodeTemplate for some of their nodes. For example, the user might decide to initialize an empty nodeTemplate for all nodes, and then only set virtualCapacities for nodes of a very specific type.
2. It is no longer valid, I corrected it to only specify `virtualCapacities`, thank you.
3. It is no longer valid, I removed that from the description, thanks.

Maydaer commented on pkg/apis/aws/validation/worker.go:
```diff
@@ -83,6 +81,11 @@ func ValidateWorkerConfig(workerConfig *apisaws.WorkerConfig, volume *core.Volum
 		}
+
+		for capacityAttribute, value := range nodeTemplate.VirtualCapacity {
+			// extended resources are required to be whole, non-negative numbers https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#consuming-extended-resources
```
Thanks, I updated the comment. I think I mixed it up with the german word for integers.

aaronfern commented:
> I don't see an important use case for allowing an empty nodeTemplate, but I also do think it is in principle valid to have an empty nodeTemplate. The behavior for an empty nodeTemplate is well defined. I think you can argue that the user might have forgotten to set a capacity or virtualCapacity if both are empty, but the user might also have intentionally chosen an implementation that has the potential to enter an empty nodeTemplate for some of their nodes. For example, the user might decide to initialize an empty nodeTemplate for all nodes, and then only set virtualCapacities for nodes of a very specific type.
Edit: I will change the check to require at least one of capacity or virtualCapacity to be non-empty, if you think that is correct.

I don't really have a strong opinion here. I'm fine with having an empty nodeTemplate if there is some value to it.
So, as of now this PR seems fine to me, but maybe @kon-angelo or @elankath could weigh in on this point
